graphql/src/main/java/com/linecorp/armeria/server/graphql/DefaultGraphQLService.java

```diff
@@ -0,0 +1,201 @@
+        this(graphQL, null);
+    }
+
+    DefaultGraphQLService(GraphQL graphQL, @Nullable DataLoaderRegistry dataLoaderRegistry) {
+        this.graphQL = graphQL;
+        this.dataLoaderRegistry = dataLoaderRegistry;
+    }
+
+    @Override
+    protected HttpResponse doGet(ServiceRequestContext ctx, HttpRequest request) {
+        final QueryParams queryString = QueryParams.fromQueryString(ctx.query());
+        final String query = queryString.get("query");
+        if (query == null) {
+            return HttpResponse.of(HttpStatus.BAD_REQUEST,
+                                   MediaType.PLAIN_TEXT,
+                                   "Query is required");
+        }
+        final Map<String, Object> variables = toVariableMap(queryString.get("variables"));
+        final String operationName = queryString.get("operationName");
+        return execute(executionInput(query, ctx, variables, operationName));
+    }
+
+    @Override
+    protected HttpResponse doPost(ServiceRequestContext ctx, HttpRequest request) {
+        final MediaType contentType = request.contentType();
+        if (contentType == null) {
+            return HttpResponse.of(HttpStatus.UNPROCESSABLE_ENTITY,
+                                   MediaType.PLAIN_TEXT,
+                                   "Could not process GraphQL request");
+        }
+        if (contentType.is(MediaType.JSON) || contentType.subtype().endsWith("+json")) {
+            return HttpResponse.from(request.aggregate().handleAsync((req, thrown) -> {
+                if (thrown != null) {
+                    logger.warn("{} Failed to aggregate a request:", ctx, thrown);
+                    return HttpResponse.ofFailure(thrown);
+                }
+                final String body = req.contentUtf8();
+                if (Strings.isNullOrEmpty(body)) {
+                    return HttpResponse.of(HttpStatus.BAD_REQUEST, MediaType.PLAIN_TEXT, "Body is required");
+                }
+                final Map<String, Object> requestMap = parseJsonString(body);
+                final String query = (String) requestMap.getOrDefault("query", "");
+                return execute(executionInput(query, ctx,
+                                              toVariableMap(requestMap.get("variables")),
+                                              (String) requestMap.get("operationName")));
+            }));
+        }
+        if (contentType.is(MediaType.GRAPHQL)) {
+            return HttpResponse.from(request.aggregate().handleAsync((req, thrown) -> {
+                if (thrown != null) {
+                    logger.warn("{} Failed to aggregate a request:", ctx, thrown);
+                    return HttpResponse.ofFailure(thrown);
+                }
+                final String query = req.contentUtf8();
+                return execute(executionInput(query, ctx, null, null));
+            }));
+        }
+        return HttpResponse.of(HttpStatus.UNPROCESSABLE_ENTITY,
```

Comment:
Question: Is this the spec? Just wanted to know. 😄

Reply:
It is not a spec. Would it be handled in a different way?